source/base/mpi_remote_point_evaluation.cc

```diff
@@ -0,0 +1,200 @@
+#include <deal.II/grid/tria.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace Utilities
+{
+  namespace MPI
+  {
+    template <int dim, int spacedim>
+    RemotePointEvaluation<dim, spacedim>::RemotePointEvaluation(
+      const double tolerance)
+      : tolerance(tolerance)
+      , ready_flag(false)
+    {}
+
+
+
+    template <int dim, int spacedim>
+    RemotePointEvaluation<dim, spacedim>::~RemotePointEvaluation()
+    {
+      if (tria_signal.connected())
+        tria_signal.disconnect();
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    RemotePointEvaluation<dim, spacedim>::reinit(
+      const std::vector<Point<spacedim>> &points,
+      const Triangulation<dim, spacedim> &tria,
+      const Mapping<dim, spacedim> &      mapping)
+    {
+#ifndef DEAL_II_WITH_MPI
+      Assert(false, ExcNeedsMPI());
+      (void)points;
+      (void)tria;
+      (void)mapping;
+#else
+      if (tria_signal.connected())
+        tria_signal.disconnect();
+
+      tria_signal =
+        tria.signals.any_change.connect([&]() { this->ready_flag = false; });
+
+      this->tria    = &tria;
+      this->mapping = &mapping;
+
+      std::vector<BoundingBox<spacedim>> local_boxes;
+      for (const auto &cell : tria.active_cell_iterators())
+        if (cell->is_locally_owned())
+          local_boxes.push_back(mapping.get_bounding_box(cell));
+
+      // create r-tree of bounding boxes
+      const auto local_tree = pack_rtree(local_boxes);
+
+      // compress r-tree to a minimal set of bounding boxes
+      const auto local_reduced_box = extract_rtree_level(local_tree, 0);
+
+      // gather bounding boxes of other processes
+      const auto global_bboxes =
+        Utilities::MPI::all_gather(tria.get_communicator(), local_reduced_box);
+
+      const GridTools::Cache<dim, spacedim> cache(tria, mapping);
+
+      const auto data =
+        GridTools::internal::distributed_compute_point_locations(
+          cache, points, global_bboxes, tolerance, true);
+
+      this->recv_ranks = data.recv_ranks;
+      this->recv_ptrs  = data.recv_ptrs;
+
+      this->send_ranks = data.send_ranks;
+      this->send_ptrs  = data.send_ptrs;
+
+      this->recv_permutation = {};
+      this->recv_permutation.resize(data.recv_components.size());
+      this->point_ptrs.assign(points.size() + 1, 0);
+      for (unsigned int i = 0; i < data.recv_components.size(); ++i)
+        {
+          AssertIndexRange(std::get<2>(data.recv_components[i]),
+                           this->recv_permutation.size());
+          this->recv_permutation[std::get<2>(data.recv_components[i])] = i;
+
+          AssertIndexRange(std::get<1>(data.recv_components[i]) + 1,
+                           this->point_ptrs.size());
+          this->point_ptrs[std::get<1>(data.recv_components[i]) + 1]++;
+        }
+
+      unique_mapping = true;
+      for (unsigned int i = 0; i < points.size(); ++i)
+        {
+          unique_mapping &= this->point_ptrs[i + 1] == 1;
```

Comment:
I find this pretty hard to read. Why not just
```suggestion
          if (unqiue_mapping && this->point_ptrs[i + 1] != 1)
            unique_mapping = false;
```
?